lib/charms/data_platform_libs/v0/database_provides.py

```diff
@@ -0,0 +1,293 @@
+
+    def __init__(self, charm: CharmBase, relation_name: str = "database") -> None:
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation = self.charm.model.get_relation(relation_name)
+        self.framework.observe(
+            charm.on[relation_name].relation_changed,
+            self._on_database_relation_changed,
+        )
+
+    def _diff(self, event: RelationChangedEvent) -> dict:
+        """Retrieves the diff of the data in the relation changed databag.
+
+        Args:
+            event: relation changed event.
+
+        Returns:
+            a dict containing the added, deleted and changed
+                keys from the event relation databag.
+        """
+        # Retrieve the old data from the data key in the application relation databag.
+        old_data = json.loads(self.relation.data[self.charm.model.app].get("data", "{}"))
+        # Retrieve the new data from the event relation databag.
+        new_data = event.relation.data[event.app]
+
+        # These are the keys that were added to the databag and triggered this event.
+        added = new_data.keys() - old_data.keys()
+        # These are the keys that were removed from the databag and triggered this event.
+        deleted = old_data.keys() - new_data.keys()
+        # These are the keys that already existed in the databag,
+        # but had their values changed.
+        changed = {
+            key for key in old_data.keys() & new_data.keys() if old_data[key] != new_data[key]
+        }
+
+        # TODO: evaluate the possibility of losing the diff if some error
+        # happens in the charm before the diff is completely checked (DPE-412).
+        # Convert the new_data to a serializable format and save it for a next diff check.
+        data = {key: value for key, value in new_data.items() if key != "data"}
+        self._update_relation_data("data", json.dumps(data))
+
+        # Return the diff with all possible changes.
+        return {
+            "added": added,
+            "changed": changed,
+            "deleted": deleted,
+        }
+
+    def _on_database_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Event emitted when the database relation has changed."""
+        # Validate that the expected data has changed to emit the custom event.
+        diff = self._diff(event)
+
+        # Emit a database requested event if the database name
+        # was added to the relation databag by the application.
+        if "database" in diff["added"]:
+            self.on.database_requested.emit(event.relation)
+
+        # Emit an extra user roles requested event if the application
+        # sent some roles or updated them through the relation databag.
+        if "extra-user-roles" in diff["added"] or "extra-user-roles" in diff["changed"]:
+            self.on.extra_user_roles_requested.emit(event.relation)
+
+    def _get_relation_data(self, key: str) -> str:
```

Comment:
Given that this can return `None`, this annotation needs to be changed, or you could raise an exception here.

```suggestion
    def _get_relation_data(self, key: str) -> Optional[str]:
```

Reply:
Fixed on https://github.com/canonical/data-platform-libs/pull/2/commits/b51cb789ef94f9741bb2c8190b2fda64028f10db.